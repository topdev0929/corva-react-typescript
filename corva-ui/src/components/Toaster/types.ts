import { ReactNode } from 'react';

export type ToasterType = { message: ReactNode; autoHideDuration: number };
