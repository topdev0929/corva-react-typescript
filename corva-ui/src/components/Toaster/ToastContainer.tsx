import { useState, useEffect, ReactElement } from 'react';

import { Toaster } from './Toaster';
import { ToasterType } from './types';

type createToast = (message: ReactElement, autoHideDuration?: number) => void;

export const DEVCENTER_TOASTS_SYMBOL = Symbol.for('devcenterToasts');

export const ToastContainer = () => {
  const [toaster, setToaster] = useState<ToasterType>(null);

  const removeToast = () => setToaster(null);

  const createToast: createToast = (message, autoHideDuration) =>
    setToaster({ message, autoHideDuration });

  const showToast = createToast;

  useEffect(() => {
    window[DEVCENTER_TOASTS_SYMBOL] = {
      createToast,
      removeToast,
      showToast,
    };
  }, []);

  if (!toaster) return null;

  return (
    <Toaster
      message={toaster.message}
      autoHideDuration={toaster.autoHideDuration}
      onDismiss={removeToast}
    />
  );
};
