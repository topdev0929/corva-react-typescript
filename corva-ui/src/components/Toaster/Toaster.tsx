import { useEffect } from 'react';
import { makeStyles, Typography } from '@material-ui/core';
import { Check as CheckIcon } from '@material-ui/icons';

import { ToasterType } from './types';

const useStyles = makeStyles(theme => ({
  notificationContainer: {
    position: 'absolute',
    height: 24,
    left: 0,
    right: 0,
    display: 'flex',
    justifyContent: 'center',
    marginTop: 16,
    zIndex: 999,
  },
  notificationContent: {
    display: 'flex',
    alignItems: 'center',
    background: 'rgba(65, 65, 65, 0.7)',
    boxShadow: '0px 4px 4px rgba(0, 0, 0, 0.12)',
    borderRadius: 4,
  },
  notificationIcon: {
    width: 24,
    color: theme.palette.primary.text6,
    fontSize: 12,
    lineHeight: '24px',
  },
  notificationLabel: {
    color: theme.palette.primary.text6,
    fontWeight: 400,
    fontSize: 12,
    lineHeight: '17px',
    letterSpacing: 0.4,
    paddingRight: 12,
  },
}));

const DEFAULT_AUTO_HIDE_DURATION = 3000;

interface ToasterProps extends ToasterType {
  onDismiss: () => void;
}

export function Toaster({
  message,
  autoHideDuration = DEFAULT_AUTO_HIDE_DURATION,
  onDismiss,
}: ToasterProps): JSX.Element {
  const classes = useStyles();

  useEffect(() => {
    const timeout = setTimeout(onDismiss, autoHideDuration);

    return () => clearTimeout(timeout);
  }, []);

  return (
    <div className={classes.notificationContainer}>
      <div className={classes.notificationContent}>
        <CheckIcon className={classes.notificationIcon} />
        <Typography className={classes.notificationLabel}>{message}</Typography>
      </div>
    </div>
  );
}
