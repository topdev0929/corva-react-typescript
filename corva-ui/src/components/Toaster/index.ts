export { Toaster } from './Toaster';
export { ToastContainer } from './ToastContainer';
