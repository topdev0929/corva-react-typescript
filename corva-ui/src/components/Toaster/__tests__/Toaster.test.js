import { render, waitFor } from '@testing-library/react';

import { Toaster } from '../Toaster';

const onDismissMock = jest.fn();
const MESSAGE = 'Toaster message';

describe('Toaster', () => {
  describe('snapshots', () => {
    it('matches default snapshot', () => {
      const container = render(<Toaster message={MESSAGE} />).baseElement;
      expect(container).toMatchSnapshot();
    });
  });

  it('renders message', async () => {
    const { getByText } = render(<Toaster message={MESSAGE} onDismiss={onDismissMock} />);
    expect(getByText(MESSAGE)).toBeInTheDocument();
  });

  it('calls onDismiss', async () => {
    render(<Toaster message={MESSAGE} onDismiss={onDismissMock} />);

    await waitFor(() => expect(onDismissMock).toHaveBeenCalledTimes(1));
  });
});
