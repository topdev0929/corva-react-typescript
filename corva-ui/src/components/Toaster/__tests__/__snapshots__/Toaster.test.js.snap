// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toaster snapshots matches default snapshot 1`] = `
<body>
  <div>
    <div
      class="makeStyles-notificationContainer-1"
    >
      <div
        class="makeStyles-notificationContent-2"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root makeStyles-notificationIcon-3"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
          />
        </svg>
        <p
          class="MuiTypography-root makeStyles-notificationLabel-4 MuiTypography-body1"
        >
          Toaster message
        </p>
      </div>
    </div>
  </div>
</body>
`;
